firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Protocols/datagrams.txt

```diff
@@ -0,0 +1,456 @@
+      THEN("The transform method reports invalid sequence status")
+      THEN("After the transform method is called, the value returned by the output_datagram's seq accessor method is equal to the sequence field of the input_buffer body's header")
+      THEN("The value returned by the output_datagram's length accessor method is equal to the length field of the input_buffer body's header")
+      THEN("The paylaod returned by the output_datagram's paylaod accessor method is equal to the paylaod from the input_buffer body")
+      THEN("The output_datagram's payload accessor method returns a reference to the payload given in its constructor")
+      THEN("The data in the output_datagram's payload buffer is independent of the data in input_buffer")
+
+    WHEN("A body with a complete 2-byte header, a non-empty payload buffer consistent with the length field of the header is given to the receiver")
+      THEN("The transform method reports ok status")
+      THEN("After the transform method is called, the value returned by the output_datagram's seq accessor method is equal to the sequence field of the input_buffer body's header")
+      THEN("The value returned by the output_datagram's length accessor method is equal to the length field of the input_buffer body's header")
+      THEN("The paylaod returned by the output_datagram's paylaod accessor method is equal to the paylaod from the input_buffer body")
+      THEN("The output_datagram's payload accessor method returns a reference to the payload given in its constructor")
+      THEN("The data in the output_datagram's payload buffer is independent of the data in input_buffer")
+
+Scenario: Datagram Receiver correctly reports incrementing, rollover, and resetting of expected sequence number
+
+  GIVEN("A Datagram receiver of buffer size 254 bytes and expected sequence number equal to 0") {
+
+    // Intent : we call transform multiple times until the final sequence number is 0xff
+    WHEN("transform is called each time on parseable bodies as input_buffers, whose headers have sequence numbers incrementing from 0 to 255") 
+      THEN("All of the transform methods report ok status")
+      THEN("After the transform method is called, the value returned by the output_datagram's seq accessor method is equal to the sequence field of the input_buffer body's header")
+      THEN("The paylaod returned by the output_datagram's paylaod accessor method is equal to the paylaod from the input_buffer body")
+      THEN("The final sequence value returned by output_datagram's seq accessor method is 0xff")
+
+  GIVEN("A Datagram receiver of buffer size 254 bytes and expected sequence number equal to 1")     
+  // Setup:
+  // transform called on an input_buffer where the header's seq field has value 0
+
+  // assertions carried out in GIVEN:
+  // The output_datagram's seq accessor method returns 1
```

Comment:
Is this true? If output_datagram is coming from the transform call on line 324, I'd expect the seq method to return 0, since the body's seq field has value 0

Reply:
Sorry that's a typo. it does return 0 https://github.com/pez-globo/pufferfish-software/blob/e29b5a64925a1953ac69251b0a6f25a5b7764b68/firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Protocols/Datagrams.cpp#L1329